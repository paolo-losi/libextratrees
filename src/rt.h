# ifndef RT_H
# define RT_H

# include <stdint.h>
# include <stdio.h>
# include <stdbool.h>
# include "kvec.h"

typedef kvec_t(uint32_t) uint_vec;

typedef struct {
    float *vectors;
    double *labels;
    uint32_t n_features;
    uint32_t n_samples;
} rt_problem;


rt_problem *rt_load_libsvm_file(char *fname);

void rt_print_problem(FILE *f, rt_problem *prob);

# endif 
