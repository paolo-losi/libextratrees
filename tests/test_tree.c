#include "rt.h"
#include "test.h"


void test_leaf() {
    test_header();

    rt_problem prob;
    double vectors[] = { 1., 3.,
                         2., 4.,
                         1., 6.};
    double labels[]  = { 2., 3., 4. };

    init_problem(&prob, vectors, labels);
    rt_print_problem(&prob);
}


int main() {
    test_leaf();
    return 0;
}
