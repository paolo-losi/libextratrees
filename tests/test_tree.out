>>> test: test_leaf
problem: samples=3 features=5
sample 0. label=2 features=1 4 1 4 7
sample 1. label=2 features=3 1 1 6 8
sample 2. label=2 features=2 6 1 5 0
[DEBUG]      src/tree.c:114  >>>>> split_problem. n samples: 3
[DEBUG]      src/tree.c:149  labels are constant. generating leaf node ...
>>> test: test_split
problem: samples=5 features=6
sample 0. label=2 features=1 4 1 4 8 2
sample 1. label=2 features=3 1 1 6 6 2
sample 2. label=1 features=2 6 1 5 3 2
sample 3. label=3 features=3 2 1 7 3 3
sample 4. label=1 features=0 1 1 5 1 2
[DEBUG]      src/tree.c:114  >>>>> split_problem. n samples: 5
[DEBUG]      src/tree.c:161  number of features to test: 6
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 1
[DEBUG]      src/tree.c:189  feature index: 0
[DEBUG]      src/tree.c:193  values - min: 0 max: 3
[DEBUG]      src/tree.c:209  threshold: 2.54861
[DEBUG]      src/tree.c:57   sample_idx: 0, val: 1 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 1, val: 3 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 2, val: 2 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 3, val: 3 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 0 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 1.16667
[DEBUG]      src/tree.c:224  diversity is new best
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 2
[DEBUG]      src/tree.c:189  feature index: 5
[DEBUG]      src/tree.c:193  values - min: 2 max: 3
[DEBUG]      src/tree.c:209  threshold: 2.68581
[DEBUG]      src/tree.c:57   sample_idx: 0, val: 2 -> lower
[DEBUG]      src/tree.c:57   sample_idx: 1, val: 2 -> lower
[DEBUG]      src/tree.c:57   sample_idx: 2, val: 2 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 3, val: 3 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 2 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 1
[DEBUG]      src/tree.c:224  diversity is new best
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 3
[DEBUG]      src/tree.c:189  feature index: 4
[DEBUG]      src/tree.c:193  values - min: 1 max: 8
[DEBUG]      src/tree.c:209  threshold: 1.80751
[DEBUG]      src/tree.c:60   sample_idx: 0, val: 8 -> higher
[DEBUG]      src/tree.c:60   sample_idx: 1, val: 6 -> higher
[DEBUG]      src/tree.c:60   sample_idx: 2, val: 3 -> higher
[DEBUG]      src/tree.c:60   sample_idx: 3, val: 3 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 1 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 2
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 4
[DEBUG]      src/tree.c:189  feature index: 3
[DEBUG]      src/tree.c:193  values - min: 4 max: 7
[DEBUG]      src/tree.c:209  threshold: 5.93811
[DEBUG]      src/tree.c:57   sample_idx: 0, val: 4 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 1, val: 6 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 2, val: 5 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 3, val: 7 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 5 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 1.16667
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 5
[DEBUG]      src/tree.c:189  feature index: 2
[DEBUG]      src/tree.c:193  values - min: 1 max: 1
[DEBUG]      src/tree.c:201  constant feature
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 6
[DEBUG]      src/tree.c:189  feature index: 1
[DEBUG]      src/tree.c:193  values - min: 1 max: 6
[DEBUG]      src/tree.c:209  threshold: 2.15019
[DEBUG]      src/tree.c:60   sample_idx: 0, val: 4 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 1, val: 1 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 2, val: 6 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 3, val: 2 -> lower
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 1 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 2.5
[DEBUG]      src/tree.c:247  split found. feature_idx: 5, threshold: 2.68581
[DEBUG]      src/tree.c:114  >>>>> split_problem. n samples: 1
[DEBUG]      src/tree.c:131  min size (1) reached. current sample size: 1
[DEBUG]      src/tree.c:114  >>>>> split_problem. n samples: 4
[DEBUG]      src/tree.c:161  number of features to test: 6
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 1
[DEBUG]      src/tree.c:189  feature index: 0
[DEBUG]      src/tree.c:193  values - min: 0 max: 3
[DEBUG]      src/tree.c:209  threshold: 0.92787
[DEBUG]      src/tree.c:60   sample_idx: 0, val: 1 -> higher
[DEBUG]      src/tree.c:60   sample_idx: 1, val: 3 -> higher
[DEBUG]      src/tree.c:60   sample_idx: 2, val: 2 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 0 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 0.666667
[DEBUG]      src/tree.c:224  diversity is new best
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 2
[DEBUG]      src/tree.c:189  feature index: 5
[DEBUG]      src/tree.c:193  values - min: 2 max: 2
[DEBUG]      src/tree.c:201  constant feature
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 3
[DEBUG]      src/tree.c:189  feature index: 3
[DEBUG]      src/tree.c:193  values - min: 4 max: 6
[DEBUG]      src/tree.c:209  threshold: 5.37172
[DEBUG]      src/tree.c:57   sample_idx: 0, val: 4 -> lower
[DEBUG]      src/tree.c:60   sample_idx: 1, val: 6 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 2, val: 5 -> lower
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 5 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 0.666667
[DEBUG]      src/tree.c:169  --- new loop cycle ---
[DEBUG]      src/tree.c:187  selected feature WITHOUT replacement: 4
[DEBUG]      src/tree.c:189  feature index: 4
[DEBUG]      src/tree.c:193  values - min: 1 max: 8
[DEBUG]      src/tree.c:209  threshold: 5.39517
[DEBUG]      src/tree.c:60   sample_idx: 0, val: 8 -> higher
[DEBUG]      src/tree.c:60   sample_idx: 1, val: 6 -> higher
[DEBUG]      src/tree.c:57   sample_idx: 2, val: 3 -> lower
[DEBUG]      src/tree.c:57   sample_idx: 4, val: 1 -> lower
[DEBUG]      src/tree.c:220  regr diversity: 0
[DEBUG]      src/tree.c:224  diversity is new best
[DEBUG]      src/tree.c:233  diversity == 0
[DEBUG]      src/tree.c:247  split found. feature_idx: 4, threshold: 5.39517
[DEBUG]      src/tree.c:114  >>>>> split_problem. n samples: 2
[DEBUG]      src/tree.c:149  labels are constant. generating leaf node ...
[DEBUG]      src/tree.c:114  >>>>> split_problem. n samples: 2
[DEBUG]      src/tree.c:149  labels are constant. generating leaf node ...
